### ISSUE TITLE
Reason for escapeTick in encaseP

### ISSUE THREAD

**Issue by zerobias** [fluture-js/Fluture#148]
What was the reason for using [this approach](https://github.com/fluture-js/Fluture/blob/master/src/encase-p2.js#L25)?
I use Fluture in an extremely time- and order- sensitive project. Spent a couple of hours to understand that the encaseP resolves by at least 4-12 ms and a few ticks behind the actual resolution of the request, which causes an immediate connection failure.

setTimeout cannot guarantee immediate or exactly next tick execution
https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Reasons_for_delays_longer_than_specified

**Comment by Avaq** [comment]
Hello @zerobias, my apologies for your lost time. I did not foresee your use-case when implementing this.

The reason for delaying execution of the continuations is because I must avoid Promises catching the potential exception thrown within one of the continuations. Take this example:

```js
Future.encaseP(Promise.resolve.bind(Promise), 1)
.map(x => x + 1)
.map(x => x.foo.bar)
.fork(console.error, console.log);
```

The last `map` will throw a TypeError. Without the `escapeTick`, this code would "desugar" to:

```js
Promise.resolve(1).then(
  pipe([x => x + 1, x => x.foo.bar, console.log]),
  console.error
)
```

I'm trying to illustrate that the continuations happen synchronously. If I weren't to `escapeTick`, the Promise would catch the TypeError, and subsequently reject with it, messing up control flow.

----

That said, I'm open to support your use-case. The current implementation of `escapeTick` is very naive. I see two possible routes for optimization:

We choose an appropriate *fast* scheduler based on environment, such as `process.nextTick`, `setImmediate`, or a DOM based scheduler. I chose `setTimeout` because it's so universally available, and using something else will introduce some additional maintenance cost.

Alternatively, instead of escaping the tick, we `catch` the exception before the Promise can, and only then delay throwing it:

```js
const escapeTick = continuation => value => {
  try { continuation(value) } catch (e) {
    setTimeout(() => { throw e }, 0);
  }
}
```

The latter approach might possibly get us the best performance, but it will obfuscate the origin of the exception. What do you think?

**Comment by zerobias** [comment]
Maybe scheduler can be selected in this way?
Sometimes, "try catch" may decrease performance

```js

const immediate = (function featureDetection() {
  if (typeof setImmediate === 'function')
    return setImmediate;
  return function(escaped) { setTimeout(escaped, 0) }
})();

export const escapeTick = f => function imprisoned(x){
  immediate(function escaped(){ f(x) });
}

```

**Comment by Avaq** [comment]
Have a look at #150. Performance improved to about 500 times the original performance. :)
Will release it as a patch as soon as I have your confirmation that this solves your use-case.

**Comment by zerobias** [comment]
Great, this solved my issue 👍

**Comment by Avaq** [comment]
I released the fix in `7.1.2`. :)
